Teleporter v2.0

DL6ER opened the pull request:
**By submitting this pull request, I confirm the following (please check boxes, eg [X] - no spaces) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**

## 10

---

Re-implement Teleporter with `PharData`. The advantage is that we don't have to rely on `php-zip` being installed.

This PR implements only the re-written export so far.
The import feature has still to be coded, but shouldn't be too complicated, either.

Note that this branch was lying around for almost exactly two months because I forgot that it existed.

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

Mcat12 commented on scripts/pi-hole/php/teleporter.php:
```diff
@@ -118,40 +106,43 @@ function getWildcardListContent() {
+		$continue = strtolower($name[1]) == 'tar' && strtolower($name[2]) == 'gz' ? true : false;
 		if(!$continue || !$okay) {
-			die("The file you are trying to upload is not a .zip file (filename: ".$filename.", type: ".$type."). Please try again.");
+			die("The file you are trying to upload is not a .tar.gz file (filename: ".$filename.", type: ".$type."). Please try again.");
```
Possible XSS attack vector (`$filename` and `$type` being injected into the page)

Mcat12 commented on scripts/pi-hole/php/teleporter.php:
```diff
@@ -118,40 +106,43 @@ function getWildcardListContent() {
-		$x = $zip->open($source);
-		if ($x === true) {
-			if(isset($_POST["blacklist"]))
+		$filename = "/tmp/".$filename;
```
You can make a new variable name instead of shadowing the old `$filename`

Mcat12 commented on scripts/pi-hole/php/teleporter.php:
```diff
@@ -118,40 +106,43 @@ function getWildcardListContent() {
-		if ($x === true) {
-			if(isset($_POST["blacklist"]))
+		$filename = "/tmp/".$filename;
+		if(!move_uploaded_file($_FILES["zip_file"]["tmp_name"], $filename))
```
`$source` has been assigned to `$_FILES["zip_file"]["tmp_name"]`, so you can use the former for conciseness (and below as well)

Mcat12 commented on scripts/pi-hole/php/teleporter.php:
```diff
@@ -160,24 +151,27 @@ function getWildcardListContent() {
+	$filename = $tarname.".gz";
+	// $archive_file_name = tempnam("/tmp", "Teleporter");
+	$archive_file_name = sys_get_temp_dir() ."/". $tarname;
+	// touch($archive_file_name);
```
Remove commented code if it is dead.

DL6ER commented:
Tested it in the following way:

- Added multiple domains to all three lists (white, black, wildcard)
- Exported teleporter archive
- Removed all of them afterwards
- Imported teleporter archive
- Verified that all domains are again in the correct places

DL6ER commented:
Note that this PR now depends on https://github.com/pi-hole/pi-hole/pull/1728 being merged first.

It changes the behavior of `Teleporter 2.0` to **replace** the lists on the importing machine rather than **appending** to them (which is the behavior of `Teleporter 1.0`)
